pkg/connector/client.go

```diff
@@ -0,0 +1,265 @@
+	queryConversationPayload.IncludeConversationInfo = true
+	conversationData, err := tc.client.FetchConversationContext(conversationId, &queryConversationPayload, payload.CONTEXT_FETCH_DM_CONVERSATION)
+	if err != nil {
+		return nil, err
+	}
+
+	conversations := conversationData.ConversationTimeline.Conversations
+	if len(conversations) <= 0 {
+		return nil, fmt.Errorf("failed to find conversation by id %s", string(conversationId))
+	}
+
+	conversation := conversations[conversationId]
+	users := conversationData.ConversationTimeline.Users
+
+	methods.MergeMaps(tc.userCache, users)
+
+	return tc.ConversationToChatInfo(&conversation), nil
+}
+
+func (tc *TwitterClient) GetUserInfo(_ context.Context, ghost *bridgev2.Ghost) (*bridgev2.UserInfo, error) {
+	userInfo := tc.GetUserInfoBridge(string(ghost.ID))
+	if userInfo == nil {
+		return nil, fmt.Errorf("failed to find user info in cache by id: %s", ghost.ID)
+	}
+	return userInfo, nil
+}
+
+func (tc *TwitterClient) GetCapabilities(_ context.Context, _ *bridgev2.Portal) *bridgev2.NetworkRoomCapabilities {
+	return &bridgev2.NetworkRoomCapabilities{
+		FormattedText: false,
+		UserMentions:  true,
+		RoomMentions:  false,
+
+		Edits:         true,
+		EditMaxCount:  10,
+		EditMaxAge:    15 * time.Minute,
+		Captions:      true,
+		Replies:       true,
+		Reactions:     true,
+		ReactionCount: 1,
+	}
+}
+
+func (tc *TwitterClient) convertEditToMatrix(ctx context.Context, portal *bridgev2.Portal, intent bridgev2.MatrixAPI, existing []*database.Message, data *event.XEventMessage) (*bridgev2.ConvertedEdit, error) {
+	data.Text = strings.TrimPrefix(data.Text, "Edited: ")
+	converted, err := tc.convertToMatrix(ctx, portal, intent, data)
+	if err != nil {
+		return nil, err
+	}
+	return &bridgev2.ConvertedEdit{
+		ModifiedParts: []*bridgev2.ConvertedEditPart{converted.Parts[0].ToEditPart(existing[0])},
+	}, nil
+}
+
+func (tc *TwitterClient) convertToMatrix(ctx context.Context, portal *bridgev2.Portal, intent bridgev2.MatrixAPI, msg *event.XEventMessage) (*bridgev2.ConvertedMessage, error) {
+	partId := networkid.PartID("")
+	var MessageOptionalPartID *networkid.MessageOptionalPartID
+	if msg.ReplyData.ID != "" {
+		MessageOptionalPartID = &networkid.MessageOptionalPartID{
+			MessageID: networkid.MessageID(msg.ReplyData.ID),
+			PartID:    &partId,
+		}
+	}
+
+	textPart := &bridgev2.ConvertedMessagePart{
+		ID:   partId,
+		Type: bridgeEvt.EventMessage,
+		Content: &bridgeEvt.MessageEventContent{
+			MsgType: bridgeEvt.MsgText,
+			Body:    msg.Text,
+		},
+	}
+
+	parts := make([]*bridgev2.ConvertedMessagePart, 0)
+
+	if msg.Attachment != nil {
+		convertedAttachmentPart, indices, err := tc.TwitterAttachmentToMatrix(ctx, portal, intent, msg.Attachment)
+		if err != nil {
+			return nil, err
+		}
+		parts = append(parts, convertedAttachmentPart)
+
+		RemoveEntityLinkFromText(textPart, indices)
+	}
+
+	if len(textPart.Content.Body) > 0 {
+		parts = append(parts, textPart)
+	}
```

Comment:
If there can only be one attachment and one text part in a message, then they should always be merged (either via `cm.MergeCaption()` or just by creating them as one part in the first place). Would recommend using an empty string as the part ID always too if there's only one part (MergeCaption will use the text part's ID by default)